File: Arrays_Strings/go/string_compression.go
package main
import (
  "fmt"
  "strconv"
)


func compress(s string) string {
  var newString string
  currentLetter := string(s[0])
  var count int
  for i := 0; i < len(s); i++ {
    if string(s[i]) == currentLetter {
      count++
    } else {
      newString += currentLetter
      newString += strconv.Itoa(count)

      currentLetter = string(s[i])
      count = 1
    }
  }

  newString += currentLetter
  newString += strconv.Itoa(count)

  if len(newString) < len(s) {
    return newString
  }

  return s
}


func main() {
  str := "aabbbaccc"

  fmt.Println(compress(str))
}

File: Arrays_Strings/python/string_compression.py
def compress(string):
    new_string = ""
    current_letter = string[0]
    count = 0

    for letter in string:
        if letter == current_letter:
            count += 1
        else:
            new_string += current_letter + str(count)

            current_letter = letter
            count = 1

    new_string += letter + str(count)

    if len(new_string) < len(string):
        return new_string
    return string


def main():
    string = "aabbbaccc"
    print(compress(string))


if __name__ == '__main__':
    main()
